csharp/src/Google.Protobuf.Test/Reflection/DescriptorDeclarationTest.cs

```diff
@@ -96,76 +102,92 @@ public void MessageLocations()
         [Test]
         public void EnumComments()
         {
-            var descriptor = unitTestProto3Descriptor.FindTypeByName<EnumDescriptor>("CommentEnum");
-            Assert.NotNull(descriptor.Declaration);
-            Assert.AreEqual(" Leading enum comment\n", descriptor.Declaration.LeadingComments);
+            EnumDescriptor? descriptor = unitTestProto3Descriptor.FindTypeByName<EnumDescriptor>("CommentEnum");
+            Assert.NotNull(descriptor);
+            Assert.NotNull(descriptor!.Declaration);
+            Assert.AreEqual(" Leading enum comment\n", descriptor.Declaration!.LeadingComments);
         }
 
         [Test]
         public void NestedMessageComments()
         {
-            var outer = unitTestProto3Descriptor.FindTypeByName<MessageDescriptor>("CommentMessage");
-            var nested = outer.FindDescriptor<MessageDescriptor>("NestedCommentMessage");
-            Assert.NotNull(nested.Declaration);
-            Assert.AreEqual(" Leading nested message comment\n", nested.Declaration.LeadingComments);
+            MessageDescriptor? outer = unitTestProto3Descriptor.FindTypeByName<MessageDescriptor>("CommentMessage");
```

Comment:
Here and below, I don't think we need all of these Assert.NotNull - if the test fails with a NullReferenceException, that's easy enough to track down.

Reply:
Addition of Assert.NotNull in many places is a signal to the analyzer that beyond that point ! overrides should not be needed. If the analyzers in .NET 6 were solely in use it would work, as the code analysis attribute hints like [NotNullWhen()] are taken into account. However for this case - and all of the notes you made above about how ! instances are not needed - they are needed at the moment because of the presence of old .NET versions whose compilers and analyzers are not aware of the hints.

Reply:
Aha, I see - thanks for that additional context.

(At this point I'm unlikely to get to reviewing the prod code this week, but it's on my list for next week... sorry about this delay.)